### PR TITLE
Add support for reaping Azure subscriptions

### PR DESCRIPTION
**Pull request by infinitewarp**
https://github.com/cloudigrade/cloudigrade/issues/1364

### CODE REVIEW

File: tests/test_azure_power_off_vms.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+        f"/resourceGroups/{resource_group if resource_group else uuid.uuid4()}"
+        f"/providers/{resource_provider if resource_provider else 'Microsoft.Compute'}"
+        f"/{resource_type if resource_type else 'virtualMachines'}"
+        f"/{resource_id if resource_id else uuid.uuid4()}"
+    )
+
+
+def synthesize_tags(with_bypass_tag=True):
+    """Synthesize a set of pseudorandom tags for a resource."""
+    conditional_tag_key = (
```
Comment (abellotti) [inline tests/test_azure_power_off_vms.py:31]: is this used ?  I don't see it being returned.

Comment (infinitewarp) [reply tests/test_azure_power_off_vms.py:31]: It's used as the key a few lines later, currently on line 38.

File: reaper/azure_power_off_vms.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    # We can just discard the result and assume it completes.
+    compute_client.virtual_machine_scale_set_vms.begin_power_off(
+        resource_group_name=resource_group,
+        vm_scale_set_name=scale_set.name,
+        instance_id=vm.instance_id,
+    )
+
+
+def handle_scale_set_vms(compute_client):
+    """Indentify and conditionally power off VM scale set VMs."""
```
[the remark this answers is not in the document]
Comment (infinitewarp) [reply reaper/azure_power_off_vms.py]: Good catch! I just amended and pushed.

File: tests/test_azure_power_off_vms.py (hunk 1 above)
Comment (abellotti) [reply tests/test_azure_power_off_vms.py:31]: Ahh never mind, you're right,  I was thinking  {  "conditional_tag_key": ... }  and not { conditional_tag_key: ... }

**Comment by infinitewarp** [comment]
In my Azure tenant, I had:

* `brad-rhdir2-useast-zone1-vm2-use-existing-RHsub-20220815` 1 regular VM not running
* `brad-rhdir2-useast-zone1-vm1-20220815` 1 regular VM running
* `brad-rhdir2-useast-zone1-vm1-20220922` 1 regular VM running with the bypass tag
* `brad-rhdir2-useast-scaleset-20221104` 1 scale set with 2 running VMs
* `brad-rhdir2-ussouthcentral-scaleset-20221103` 1 scale set with the bypass tag and 2 running VMs

When I run the reaper for Azure, I expect:

* `brad-rhdir2-useast-zone1-vm2-use-existing-RHsub-20220815` will be ignored
* `brad-rhdir2-useast-zone1-vm1-20220815` will be powered off
* `brad-rhdir2-useast-zone1-vm1-20220922` will be found but skipped
* `brad-rhdir2-useast-scaleset-20221104` will be powered off
* `brad-rhdir2-ussouthcentral-scaleset-20221103` will be found but skipped

Actual results when running locally confirmed these expectations:

```
❯ AZURE_TENANT_ID="${AZURE_TENANT_ID}" \
AZURE_SUBSCRIPTION_ID="${AZURE_SUBSCRIPTION_ID}" \
AZURE_CLIENT_ID="${AZURE_CLIENT_ID}" \
AZURE_CLIENT_SECRET="${AZURE_CLIENT_SECRET}" \
REAP_BYPASS_TAG="${REAP_BYPASS_TAG}" \
poetry run python3 -m reaper.azure_power_off_vms

2022-11-08T14:52:23-0500 | INFO | Preparing to power off Azure VMs.
2022-11-08T14:52:24-0500 | INFO | Found running VM brad-rhdir2-useast-zone1-vm1-20220815
2022-11-08T14:52:24-0500 | INFO | Attempting to power off VM brad-rhdir2-useast-zone1-vm1-20220815
2022-11-08T14:52:24-0500 | INFO | VM brad-rhdir2-useast-zone1-vm1-20220922 has bypass tag and will not be powered off.
2022-11-08T14:52:24-0500 | INFO | Found running VM brad-rhdir2-useast-scaleset-20221104_0
2022-11-08T14:52:24-0500 | INFO | Attempting to power off VM scale set VM brad-rhdir2-useast-scaleset-20221104_0
2022-11-08T14:52:24-0500 | INFO | Found running VM brad-rhdir2-useast-scaleset-20221104_1
2022-11-08T14:52:24-0500 | INFO | Attempting to power off VM scale set VM brad-rhdir2-useast-scaleset-20221104_1
2022-11-08T14:52:24-0500 | INFO | VM scale set brad-rhdir2-ussouthcentral-scaleset-20221103 has bypass tag and will not be powered off.
```

Running a second time finds only the bypass-tagged VMs and VMSSes, and they continue to be skipped:

```
❯ AZURE_TENANT_ID="${AZURE_TENANT_ID}" \
AZURE_SUBSCRIPTION_ID="${AZURE_SUBSCRIPTION_ID}" \
AZURE_CLIENT_ID="${AZURE_CLIENT_ID}" \
AZURE_CLIENT_SECRET="${AZURE_CLIENT_SECRET}" \
REAP_BYPASS_TAG="${REAP_BYPASS_TAG}" \
poetry run python3 -m reaper.azure_power_off_vms

2022-11-08T14:54:56-0500 | INFO | Preparing to power off Azure VMs.
2022-11-08T14:54:58-0500 | INFO | VM brad-rhdir2-useast-zone1-vm1-20220922 has bypass tag and will not be powered off.
2022-11-08T14:54:58-0500 | INFO | VM scale set brad-rhdir2-ussouthcentral-scaleset-20221103 has bypass tag and will not be powered off.
```

I temporarily overrode the behavior of the GitHub actions workflow definitions to force the normally-scheduled jobs to run whenever I pushed to this branch. I restarted some VMs and VMSSes in Azure, triggered the "scheduled" jobs, and confirmed that they behaved correctly as well:

Screen Shot 2022-11-08 at 4 36 41 PM-nq8

If the GitHub secrets ever disappear or if the Azure app+keys are unable to communicate with the configured tenant+subscription, the affected jobs fail and notify via the usual means:

fail-nq8

Since this branch/PR also slightly changed the definition of the AWS jobs, I overrode their behavior and confirmed they ran successfully too:

Screen Shot 2022-11-08 at 4 44 41 PM-nq8

**Review by BrittRH — APPROVED** [review]

**Pull request merged**